### PR TITLE
Fix issue #1100. StartInfo may return incorrect information for *running* processes.

### PR DESCRIPTION
**Pull request by heXelium**
Fix issue #1100. Implemented throwing exception mechanism in a way as @davkean  suggests in his comment to this task: 
> My proposal in this case, is to throw InvalidOperationException (with a good error message) when accessing the Process.StartInfo for Processes that you didn't start yourself. This will prevent people being confused by this behavior.

### CODE REVIEW

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs [hunk 1]
```diff
@@ -482,6 +482,11 @@ public ProcessStartInfo StartInfo
             {
                 if (_startInfo == null)
                 {
+                    if (Associated && Process.GetCurrentProcessId() != Id)
```
Comment (davkean) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs]: Why do you allows devs to get the StartInfo for the current process? It also contains bogus information; environment variables is the only thing that returns something, and even then it is snapshotted and never changed once you've read it.

Comment (davkean) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs]: Actually, looking deeper, this logic looks wrong. __Associated__ returns true after you've _started_ the Process:
'''
Process process = new Process();
process.StartInfo.UseShellExecute = false;
process.StartInfo.FileName = "notepad.exe";
process.Start();

Console.WriteLine(process.StartInfo.FileName); // This now throws

```
You should only throw when you did not start the process yourself.

Comment (heXelium) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs]: You're right that environment variables is the only one thing that is have some useful meaning. So there no too much sense in returning this object. Some of the buggy information can be retrieved and fixed for example Arguments. What do you suggest also throw same exception for current process?

Please correct me if I wrong but does environment variables can be changed for running process. I found following on MSDN:
> Altering the environment variables of a child process during process creation is the only way one process can directly change the environment variables of another process. A process can never directly change the environment variables of another process that is not a child of that process.

Comment (heXelium) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs]: Yes Associated returns true when process is started and it does not matter how it was started. But that what I'm expecting. In your example it will not throw because _startInfo is not null. It was assigned on previous steps. And there no way to start process from code and not to assign this variable. 
Anyway I'm going to remove this checking and throw exception even for current process.
```csharp
if (_startInfo == null)
{
   if (Associated && Process.GetCurrentProcessId() != Id)
   {
      throw new InvalidOperationException(SR.CantGetProcessStartInfo);
   }
   _startInfo = new ProcessStartInfo();
}
```

**Comment by heXelium** [comment]
Please check out. Changed behavior in way as you expect. Also added unit tests for edge cases.

**Comment by davkean** [comment]
Behavior looks fine to me now. @stephentoub @pallavit can you take a look?

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 2]
```diff
@@ -637,5 +637,46 @@ public void Process_Environment()
...
+                Assert.Equal(ProcessName, process.StartInfo.FileName);
+            }
+
+            {
+                Process process = Process.GetCurrentProcess();
+                Assert.Throws<System.InvalidOperationException>(() => process.StartInfo);
+            }
+
+            {
+                Process process = Process.Start(ProcessName);
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: This should use CreateProcess().  We need to avoid flowing profiler-related environment variables to the spawned process.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 3]
```diff
@@ -637,5 +637,46 @@ public void Process_Environment()
...
+                process.Start();
+                Assert.Equal(ProcessName, process.StartInfo.FileName);
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: We should kill/wait for the process.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: That said... how is this test different from the first one in the function?  Or more specifically, what's the difference you're testing between the two?

**Comment by stephentoub** [comment]
The product change LGTM.  Some minor questions on the tests.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs (hunk 2 above)
Comment (heXelium) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Eventually it's same but the second test is covering implementation of Process static start method to check if StartInfo property is assigned correctly. Do you think it's redundant?

**Comment by heXelium** [comment]
>I am not sure if real is the right word here.

I took this word from the description of Associated property which is says: "Returns whether this process component is associated with a real process". I can't come up with something better.

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: I see.  What's the purpose then of the next line then?  Won't that effectively turn it into the first case, by Close'ing the handle and then refreshing/recreating everything?

Regardless, I see your goal here, but I don't think it's necessary, especially since you'll need to handle the case I mentioned about weeding out the problematic environment variables.

Comment (heXelium) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Agreed it's reasonable, removed this test.

**Comment by pallavit** [comment]
> I am not sure if real is the right word here.

To be honest, I could not come up with anything else either :|. So lets keep it as-is for now.

**Comment by pallavit** [comment]
LGTM. @stephentoub do you have any further concerns?

**Comment by stephentoub** [comment]
LGTM.

**Comment by pallavit** [comment]
@heXelium  Could you please resolve the merge conflicts. Thanks.

**Comment by heXelium** [comment]
@pallavit done with some minor changes in UT.

**Comment by stephentoub** [comment]
Thanks, @heXelium.
